File: go/src/create_new_deposit/src/create_deposit.go
package create_deposit

import (
	"strconv"

	deposit "interview_mock_deposits_go/deposit/src"
	"interview_mock_deposits_go/deposit/src/domain"
	"interview_mock_deposits_go/internal/events"

	"github.com/sirupsen/logrus"
)

func CreateDeposit(event events.PendingTransactionEvent, requestId string) error {
	logrus.WithFields(logrus.Fields{
		"transactionId": event.TransactionId,
		"customerId":    event.CustomerId,
	}).Info("creating deposit for pending transaction")

	amount, err := strconv.ParseFloat(event.Amount, 64)
	if err != nil {
		return err
	}

	transaction := deposit.Transaction{
		UserId:        event.CustomerId,
		Type:          domain.SalaryAdvance,
		TransactionId: event.TransactionId,
		Amount:        amount,
	}

	userData := deposit.UserData{
		Name:     event.Name,
		Document: event.Document,
		BankAccount: deposit.BankAccount{
			Bank:    event.BankAccount.Bank,
			Branch:  event.BankAccount.Branch,
			Account: event.BankAccount.Account,
		},
	}

	depositId, err := deposit.CreateDepositWithUserData(transaction, userData, requestId)
	if err != nil {
		return err
	}

	// mock event publication (EventBridge PutEvents)
	logrus.WithFields(logrus.Fields{
		"detailType": events.DepositRequestedType,
		"depositId":  depositId,
	}).Info("event would be published")

	return nil
}

File: python/src/create_new_deposit/create_deposit.py
import logging

from deposit import (
    BankAccount,
    Transaction,
    UserData,
    create_deposit_with_user_data,
)
from deposit.domain.transaction_type import TransactionType
from internal.events import PendingTransactionEvent

log = logging.getLogger(__name__)


def create_deposit(event: PendingTransactionEvent, request_id: str) -> None:
    log.info(
        "creating deposit for pending transaction",
        extra={"transactionId": event.transaction_id, "customerId": event.customer_id},
    )

    amount = float(event.amount)

    transaction = Transaction(
        user_id=event.customer_id,
        type=TransactionType.SALARY_ADVANCE,
        transaction_id=event.transaction_id,
        amount=amount,
    )

    user_data = UserData(
        name=event.name,
        document=event.document,
        bank_account=BankAccount(
            bank=event.bank_account.bank,
            branch=event.bank_account.branch,
            account=event.bank_account.account,
        ),
    )

    create_deposit_with_user_data(transaction, user_data, request_id)
